### PR TITLE
Configure doctr

### PR DESCRIPTION
**Pull request by thequackdaddy**
So this is my attempt to configure doctr for this project. 

A few steps might need to be taken...

1. Go to https://github.com/statsmodels/statsmodels.github.io/settings/keys
   and add the following as a new key with write access: (this is a public key, so can be posted here.)

    ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDXFzRsqaj5m8LEZ1LjIX2HhChdINtia4C7gEh4TZIJ+Dd6GYrxYV5mJlQoZcalg5kbFcfhbSaB880wxqd5WFWaFZEEhqf7fR2jkeOwK0iNzJiPuYpHvOYqcJ8kAB1v5T81jUXTxKTSuPq3N7VKwt2ZMdPoioxLY0NBiUfnhh7s8TrWrqjT9cWApGYUYpann/gjvKtoIr4pWP5VT07gYXc//6b3vcv8H/yersy9W3rP825M/8UanpOQZdgKO8Uv9F/ivgc/fPFHE9NaL3fnVEM0cDCjl9/vSpaLV9sfVTYysqWw0k19DdR2Psfrp1tWpKBpdQrayTeW04lIj54YZQXcuMGlJakwkEX2GdUfRqgBzX4eYimHhua9RhzBlQ8cWbrJrQkpIQUqHV1eN6x18nDMphJsbGKQDU7eDIsKNMtAXdg68uVwDSIshNNJ8AMVaTb2FG/REBdsHtIhtt0l03zawslz1PooHfGIgdxljaUL7YCKha0K7xQ59uyhmgLGGUpJzSEMqybUTmGHU/3pWF1geWlVcODrhK6EZ08tiQTVCPN2rU1G36JzW70xVchwPRQhX10ap7vszp+UO6MkKlEKR6MHPtY5X/bzSgS1eqe02h8KhHYtUPjv71zkF94TrCWR+UQqBg7lyH3BFat1oYEYV/qSMRgSoq35mTXrk6rcfw==

(edit) Follow rest of steps are below... 

Hopefully this is helpful.

### CODE REVIEW

**Comment by bashtage** [comment]
I don't think you need to delete devel.  I think doctr does a force push of gh-pages so that history doesn't explode.

**Comment by thequackdaddy** [comment]
@bashtage In retrospect, I think you might be right. I was trying to set it up so that the entire history was in one branch. I *think* the way gh-pages works is that it searches for gh-pages first and then falls back to master. So if you have a non-existent gh-pages, doctr will create a new, empty branch then just push there. Requests to https://statsmodels.github.io/devel will find the gh-pages branch first for the web. 

The only downside, is that gh-pages won't have the full history of all the docs, but that doesn't really matter.

**Comment by thequackdaddy** [comment]
@bashtage Looking at the travis log here: https://api.travis-ci.org/v3/job/387138982/log.txt

I see this near the end...

```
[34mgit add devel/_images/arma_predict_plot.png[0m
fatal: pathspec 'devel/_images/arma_predict_plot.png' is beyond a symbolic link
[31mERROR: The doctr command '/home/travis/miniconda/envs/statsmodels-test/bin/doctr deploy --built-docs docs/build/html/ --deploy-repo statsmodels/statsmodels.github.io devel' failed: SystemExit('\x1b[31mgit add devel/_images/arma_predict_plot.png failed: 128\x1b[0m',)[0m
[34mgit checkout 6b04b28be1c3a50bf1bc082dad7a2f6b4f7c91ed[0m
error: Your local changes to the following files would be overwritten by checkout:
	0.9.0/_images/arma_predict_plot.png
	0.9.0/_images/bkf_plot.png
	0.9.0/_images/cff_plot.png
```

Maybe I'm wrong. It looks like it is copying the `master` branch after all. But it gives an error because `devel` is a symbolic link. Maybe we should delete that to be safe? 

Or is this just because this is a PR?

**Comment by thequackdaddy** [comment]
@josef-pkt Any interest/thoughts on this?

**Comment by josef-pkt** [comment]
I would have preferred to use regular branches instead of gh-pages.

However, after (not much) thinking about it, I prefer to do whatever you and Kevin recommend.
(i.e. I don't want to have to figure out the details.)

One thing that we lost with moving from sourceforge to github is the top landing page with the intro and link to different versions.
I don't know if it will be possible to recover it when using doctr for automatically creating the entire side.
Can we link from master to gh-pages if needed, or is there automatic redirect?

Do we need to adjust the website link for the www.statsmodels.org domain?

(aside: In less than a week I will start my annual family vacation for a month in Europe, Central European Time zone)

**Comment by thequackdaddy** [comment]
> I would have preferred to use regular branches instead of gh-pages.

I think we can use the argument `--deploy-branch-name master` to account for that. 

> One thing that we lost with moving from sourceforge to github is the top landing page with the intro and link to different versions. I don't know if it will be possible to recover it when using doctr for automatically creating the entire side.

Sympy does this with doctr by having a python script that looks for all the versions in directory. See this... https://github.com/sympy/sympy_doc/blob/gh-pages/generate_indexes.py

> Can we link from master to gh-pages if needed, or is there automatic redirect?

I think, reading from gh-pages docs, that when you ask for https://statsmodels.github.io/ it first looks for the gh-pages branch in https://github.com/statsmodels/statsmodels.github.io, then goes to master if it doesn't find the requested html file, or index.html. 

I would highly suggest removing the symlinks devel and stable and copying 0.9.0 to stable if we go down this route and merge this PR. I'm not confident that doctr will handle symlinks well. 

 > Do we need to adjust the website link for the www.statsmodels.org domain?

I think the answer is "No". You need a CNAME file, which we already have.

**Comment by bashtage** [comment]
Do think this one is ready?

**Comment by thequackdaddy** [comment]
@bashtage Do I think its finished? I think its a good start. 

I imagine the hardest part will be getting it set up the first time.

**Comment by thequackdaddy** [comment]
@josef-pkt 

> One thing that we lost with moving from sourceforge to github is the top landing page with the intro and link to different versions.

I've been experimenting with this. 

See this PR #5088

I think basically does what you miss. Thoughts?

**Comment by thequackdaddy** [comment]
I wrote a quick file for statsmodels.github.io that should run to generate a file called `versions.json` which is used for a drop down menu so you can pick between other versions of the docs.

**Comment by thequackdaddy** [comment]
@josef-pkt I just rebased this again. Keep getting merge conflicts because of all the travis changes.

**Comment by bashtage** [comment]
@thequackdaddy I think the .travis.yml changes are over for now

@josef-pkt This is a really useful PR that will help a lot in documentation generation (in particular in doc building for devel, but also simplifies release doc building).  It should get in as soon as possible.

**Comment by thequackdaddy** [comment]
There is one small doc problem that is raising an exception now that we've cleaned up the `api.py` modules. Fixed (I think) in last push.

**Comment by bashtage** [comment]
It failed again but I can't see anything in the log as to why.

File: .travis.yml [hunk 1]
```diff
@@ -138,7 +140,21 @@ script:
   # Show versions
   - python -c 'import statsmodels.api as sm; sm.show_versions();'
   # docbuild and exit, if required
-  - if [ ${DOCBUILD} = true ]; then cd ${SRCDIR}/docs; source ${SRCDIR}/tools/ci/docbuild.sh; exit 0; fi;
+  - |
+    if [ ${DOCBUILD} = true ]; then
```
Comment (bashtage) [inline .travis.yml]: Should this all go in docbuild.sh along with some echo to explain where the script is?

File: .travis.yml [hunk 2]
```diff
@@ -138,7 +140,20 @@ script:
...
+      cd ${SRCDIR}/docs;
+      bash ${SRCDIR}/tools/ci/docbuild.sh;
```
Comment (bashtage) [inline .travis.yml]: Should this be source not bash?

File: .travis.yml (hunk 1 above)
Comment (thequackdaddy) [reply .travis.yml]: Clever. Yes, let me try that.

**Comment by thequackdaddy** [comment]
@bashtage I removed the `set -e` and moved some things around and I think this is a little better? Time to ask @josef-pkt to review/merge? 

Steps for @josef-pkt (or whoever wants to merge):

1. Go to https://github.com/statsmodels/statsmodels.github.io/settings/keys
and add the following as a new key with write access: (this is a public key, so can be posted here.)

ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDXFzRsqaj5m8LEZ1LjIX2HhChdINtia4C7gEh4TZIJ+Dd6GYrxYV5mJlQoZcalg5kbFcfhbSaB880wxqd5WFWaFZEEhqf7fR2jkeOwK0iNzJiPuYpHvOYqcJ8kAB1v5T81jUXTxKTSuPq3N7VKwt2ZMdPoioxLY0NBiUfnhh7s8TrWrqjT9cWApGYUYpann/gjvKtoIr4pWP5VT07gYXc//6b3vcv8H/yersy9W3rP825M/8UanpOQZdgKO8Uv9F/ivgc/fPFHE9NaL3fnVEM0cDCjl9/vSpaLV9sfVTYysqWw0k19DdR2Psfrp1tWpKBpdQrayTeW04lIj54YZQXcuMGlJakwkEX2GdUfRqgBzX4eYimHhua9RhzBlQ8cWbrJrQkpIQUqHV1eN6x18nDMphJsbGKQDU7eDIsKNMtAXdg68uVwDSIshNNJ8AMVaTb2FG/REBdsHtIhtt0l03zawslz1PooHfGIgdxljaUL7YCKha0K7xQ59uyhmgLGGUpJzSEMqybUTmGHU/3pWF1geWlVcODrhK6EZ08tiQTVCPN2rU1G36JzW70xVchwPRQhX10ap7vszp+UO6MkKlEKR6MHPtY5X/bzSgS1eqe02h8KhHYtUPjv71zkF94TrCWR+UQqBg7lyH3BFat1oYEYV/qSMRgSoq35mTXrk6rcfw==

2. Merge my PR at statsmodels/statsmodels.github.io#20
3. Merge this PR.
4. doctr will setup a new branch called gh-pages. That branch will be pushed to and all future web requests to the statsmodels.org domain will go there. You should protect that branch so that doctr doesn't accidentally blow up the history of that repository.

**Comment by josef-pkt** [comment]
I did steps 1, 2 and 3.  (easy enough even for a maintainer that doesn't pay attention.)

Let's see whether it works. It will be great when we have this working on automatic
@thequackdaddy @bashtage thank you

**Comment by bashtage** [comment]
Tune back in 20 minutes.

**Comment by bashtage** [comment]
Make that 2 hours.

**Comment by thequackdaddy** [comment]
Well it worked, but it took about 46 minutes to work. 

It looks like does a silly `git add file` to each and every file. That whole process takes about 20 minutes, I think. 

Some potential fixes/workarounds:

* Many of the files it commits are just *.txt files that (I presume) sphinx uses to stage the docs. These can be deleted.
* It (may) be possible to manually `git add docs/generated/*` so that doctr doesn't have to do that. 
* Re-do sphinx/project to create fewer files. 
* Other ideas? 

But it worked!

**Comment by bashtage** [comment]
Seems to have mostly worked:

http://www.statsmodels.org/devel/

Might need to add a git fetch --tags line to .travis before calling setup.

**Comment by bashtage** [comment]
@thequackdaddy Ideally doctr upstream would fix the git add problem in a nice way.

**Comment by thequackdaddy** [comment]
Also, I realized that it pushed to the branch `master`, not `gh-pages` as I had expected. Doesn't matter, but somewhat surprising.

**Comment by bashtage** [comment]
Looks like --command <command> might do the trick.

```
--command <command>¶
Command to be run before committing and pushing. This command will be run from the deploy repository/branch. If the command creates additional files that should be deployed, they should be added to the index.

--no-sync
Don’t sync any files. This is generally used in conjunction with the –command flag, for instance, if the command syncs the files for you. Any files you wish to commit should be added to the index.
```

**Comment by thequackdaddy** [comment]
@bashtage 

> Ideally doctr upstream would fix the git add problem in a nice way.

Fair point. Do you want to open an issue there? I took a look at their code and I think this is done here:

https://github.com/drdoctr/doctr/blob/5067c1e28ef446ef058157bce2441554a2710588/doctr/travis.py#L477-L482

Clearly, it would be nice you could just `git add devel/*` or whatever.

**Comment by thequackdaddy** [comment]
@bashtage 

As for the `--command` comment, yes, that's what I was thinking. I think a simple "git add docs/generated/*"? (or whatever the folders are)

**Pull request merged**